### PR TITLE
Controls (Regen)

### PR DESCRIPTION
**Pull request by daniwhite**
- Regen braking
- Also some old controls tests

### CODE REVIEW

File: can_spec_my18.yml [hunk 1]
```diff
@@ -531,22 +555,31 @@ buses:
...
-              TRACTION_CONTROL_DISABLE: 0x6
-              ACTIVE_AERO_ENABLE: 0x7
-              ACTIVE_AERO_DISABLE: 0x8
-              FAN_ENABLE: 0x9
-              FAN_DISABLE: 0xA
+            length: 1
+            c_type: bool
+          regen_bias: # Ranges from 0 to 100
+            position: 1
+            length: 16
```
Comment (nistath) [inline can_spec_my18.yml:565]: Too big?

File: can_spec_my18.yml [hunk 2]
```diff
@@ -531,22 +555,31 @@ buses:
             c_type: int32_t
             signed: false
             unit: Wh
-      DashRequest:
+      DashControls:
```
Comment (nistath) [inline can_spec_my18.yml:554]: I'd like to bunch up all future, potentially non-controls related (if any), dash requests in one message, so make sure to keep everything as small as can be.

File: src/vcu/inc/state_vcu.h [hunk 3]
```diff
@@ -46,14 +49,26 @@ typedef struct {
   bool ScrollSelect;
 } Buttons_T;
 
+typedef struct {
+  int16_t north;
+  int16_t east;
+} IMUVelocity_T;
```
Comment (nistath) [inline src/vcu/inc/state_vcu.h:55]: It's a good practice to name these state structs by what they contain and not where they get the information from.

File: src/vcu/inc/state_vcu.h [hunk 4]
```diff
@@ -46,14 +49,26 @@ typedef struct {
...
+
+typedef struct {
+  uint32_t front_left_32b_wheel_speed;
+  uint32_t front_left_16b_wheel_speed;
+  uint32_t front_right_32b_wheel_speed;
+  uint32_t front_right_16b_wheel_speed;
```
Comment (nistath) [inline src/vcu/inc/state_vcu.h:61]: Unnecessary `_wheel_speed` suffix.

File: src/vcu/src/can_handles.c [hunk 5]
```diff
@@ -147,6 +170,69 @@ void handleButtonRequest(Frame *msg) {
...
+
+  mc_readings.torque_feedback = unpacked_msg.torque_feedback;
+}
+
+void handleDashControls(Frame *msg) {
+  can0_DashControls_T unpacked_msg;
+
+  unpack_can0_DashControls(msg, &unpacked_msg);
+
+  control_settings.using_regen;
```
Comment (nistath) [inline src/vcu/src/can_handles.c:233]: Consider a typedef that sets Controls_Settings_T to the CANlib struct and just assign the unpacked struct to the instance declared in controls.h.

File: src/vcu/src/controls.c [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        if (pedalbox_max(accel) < LC_ACCEL_RELEASE) {
+          lc_state = DONE;
+          printf("[LAUNCH CONTROL] DONE STATE ENTERED\r\n");
+        }
+        break;
+    }
+
+  } else {
+    // Control regen brake valve:
+    bool brake_valve_state = control_settings.using_regen && get_pascals(pedalbox.REAR_BRAKE) < RG_REAR_BRAKE_THRESH;
```
Comment (nistath) [inline src/vcu/src/controls.c:91]: You should also have an upper bound for front brake pressure, distress braking, which would open the valve.

File: src/vcu/inc/controls.h [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+// Launch control contants
+// After wheel speed crosses this threshold, start using slip controller
+#define LC_WS_THRESH 45000
+#define LC_cGR 347 // Gear ratio times 100
+#define LC_ACCEL_BEGIN 950 // 95%
+#define LC_ACCEL_RELEASE 50 // 5%
+#define LC_BRAKE_BEGIN 100 // We want a lower threshold
+
+// RG = regen
+#define RG_MOTOR_SPEED_THRESH 250            // RPM
```
Comment (nistath) [inline src/vcu/inc/controls.h:25]: It is good to use MIN or MAX in lieu of THRESH in order to indicate the type of bound intended.

File: src/vcu/src/controls.c [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+static int32_t get_regen_torque() {
+  int32_t regen_torque = 0;
+
+  // IMU velocity component unit conversion:
+  // Starting units: 100 * m/s, ending units: kph
+  // (100 m/s) * (1 km/1000 m) * (3600 s/1 hr) = 360 kph
+  // So multiply by 360 to get kph --> but this is squared, so multiply by 129600
+  int32_t imu_vel_north_squared = imu_velocity.north * imu_velocity.north;
+  int32_t imu_vel_east_squared = imu_velocity.east * imu_velocity.east;
+  int32_t car_speed_squared = 129600 * (imu_vel_north_squared + imu_vel_east_squared);
```
Comment (nistath) [inline src/vcu/src/controls.c:128]: Why are you squaring?

File: src/vcu/src/controls.c [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+  if ((mc_readings.speed * -1 > RG_MOTOR_SPEED_THRESH) &&
+      (cs_readings.V_bus < RG_BATTERY_VOLTAGE_MAX_THRESH) &&
+      (get_pascals(pedalbox.FRONT_BRAKE) > RG_FRONT_BRAKE_THRESH) &&
+      car_speed_squared > 25) { // Threshold is 5 kph, so 25 (kph)^2
```
Comment (nistath) [inline src/vcu/src/controls.c:133]: You defined a threshold for this in `controls.h`.

File: src/vcu/src/controls.c [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+  // IMU velocity component unit conversion:
+  // Starting units: 100 * m/s, ending units: kph
+  // (100 m/s) * (1 km/1000 m) * (3600 s/1 hr) = 360 kph
+  // So multiply by 360 to get kph --> but this is squared, so multiply by 129600
+  int32_t imu_vel_north_squared = imu_velocity.north * imu_velocity.north;
+  int32_t imu_vel_east_squared = imu_velocity.east * imu_velocity.east;
+  int32_t car_speed_squared = 129600 * (imu_vel_north_squared + imu_vel_east_squared);
+
+  if ((mc_readings.speed * -1 > RG_MOTOR_SPEED_THRESH) &&
+      (cs_readings.V_bus < RG_BATTERY_VOLTAGE_MAX_THRESH) &&
```
Comment (nistath) [inline src/vcu/src/controls.c:131]: Use the motor controller voltage until we find out if the current sensor is okay.

File: src/vcu/src/controls.c [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    // pascals, we only need to divde by 10^4, not 10^7 for RG_10_7_K
+    // And by dividing by 10^3 instead of 10^4, we are multiplying by 10 to get
+    // dNm instead of Nm
+    regen_torque = RG_10_7_K * kilo_pascals * (100 - control_settings.cBB_ef) /(control_settings.cBB_ef * 1000);
+    if (regen_torque > RG_TORQUE_COMMAND_MAX) {
+      regen_torque = RG_TORQUE_COMMAND_MAX;
+    }
+  }
+
+  return -1 * regen_torque;
```
Comment (nistath) [inline src/vcu/src/controls.c:145]: Why is this negative?

File: src/vcu/src/controls.c [hunk 12]
```diff
@@ -0,0 +1,191 @@
+#include "controls.h"
+
+static bool enabled = false;
+static int16_t torque_command = 0;
+static int16_t speed_command = 0;
+Controls_Settings_T control_settings = {};
```
Comment (nistath) [inline src/vcu/src/controls.c:6]: Should be `volatile`. I doubt it would ever be optimized away, but it should be done.

**Review by nistath — APPROVED** [review]
LGTM

File: can_spec_my18.yml (hunk 1 above)
Comment (daniwhite) [reply can_spec_my18.yml:565]: Yeah, I was on the fence about this...but I was thinking about our conversation the other day about how it benefit vs. cost of using a larger integer size

Comment (daniwhite) [reply can_spec_my18.yml:565]: Given your other comment I suppose I can squish them

File: src/vcu/src/controls.c (hunk 8 above)
Comment (daniwhite) [reply src/vcu/src/controls.c:128]: Because I want the magnitude and have have components

File: src/vcu/src/controls.c (hunk 10 above)
Comment (daniwhite) [reply src/vcu/src/controls.c:131]: Current sensor is okay (or I have never seen it not be okay)
Please show me evidence of it not being okay

File: src/vcu/src/controls.c (hunk 11 above)
Comment (daniwhite) [reply src/vcu/src/controls.c:145]: Because regen is negative torque
I suppose I can add a comment

File: src/vcu/src/controls.c (hunk 12 above)
Comment (daniwhite) [reply src/vcu/src/controls.c:6]: :+1:

File: src/vcu/src/controls.c (hunk 8 above)
Comment (daniwhite) [reply src/vcu/src/controls.c:128]: And I don't wanna take square root (is there a good way to do this?)

File: src/vcu/src/controls.c (hunk 9 above)
Comment (daniwhite) [reply src/vcu/src/controls.c:133]: Oh sorry I'll fix that

File: src/vcu/src/controls.c (hunk 8 above)
Comment (daniwhite) [reply src/vcu/src/controls.c:128]: Yeah I later do wrong math in the comparison

File: src/vcu/src/controls.c (hunk 10 above)
Comment (nistath) [reply src/vcu/src/controls.c:131]: It feels ok

File: src/vcu/src/controls.c (hunk 11 above)
Comment (nistath) [reply src/vcu/src/controls.c:145]: oh pls thx

File: can_spec_my18.yml (hunk 1 above)
Comment (nistath) [reply can_spec_my18.yml:565]: Don't squish unless you plan on reflashing dash. Maybe squish if and when we need it.

File: src/vcu/src/controls.c (hunk 12 above)
Comment (skoppula) [reply src/vcu/src/controls.c:6]: :+1:

**Pull request merged**